File: lib/rag/api/repository.ts
/**
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import { Duration } from 'aws-cdk-lib';
import { IAuthorizer, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { ISecurityGroup } from 'aws-cdk-lib/aws-ec2';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { IFunction, ILayerVersion } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

import { getDefaultRuntime, PythonLambdaFunction, registerAPIEndpoint } from '../../api-base/utils';
import { BaseProps } from '../../schema';
import { Vpc } from '../../networking/vpc';
import { LAMBDA_PATH } from '../../util';

/**
 * Properties for RepositoryAPI Construct.
 *
 * @property {IAuthorizer} authorizer - APIGW authorizer
 * @property {Record<string,string>} baseEnvironment - Default environment properties applied to all
 *                                                      lambdas
 * @property {ILayerVersion[]} commonLayers - Lambda layers for all Lambdas.
 * @property {IRole} lambdaExecutionRole - Execution role for lambdas
 * @property {string} restApiId - REST APIGW for UI and Lambdas
 * @property {ISecurityGroup[]} securityGroups - Security groups for Lambdas
 * @property {Vpc} vpc - Stack VPC
 */
type RepositoryApiProps = {
    authorizer?: IAuthorizer;
    baseEnvironment: Record<string, string>;
    commonLayers: ILayerVersion[];
    lambdaExecutionRole: IRole;
    restApiId: string;
    rootResourceId: string;
    securityGroups: ISecurityGroup[];
    vpc: Vpc;
} & BaseProps;

/**
 * API for RAG repository operations
 */
export class RepositoryApi extends Construct {
    public createCollectionFunction: IFunction;

    constructor (scope: Construct, id: string, props: RepositoryApiProps) {
        super(scope, id);

        const {
            authorizer,
            baseEnvironment,
            commonLayers,
            lambdaExecutionRole,
            restApiId,
            rootResourceId,
            securityGroups,
            vpc,
            config
        } = props;

        const restApi = RestApi.fromRestApiAttributes(this, 'RestApi', {
            restApiId: restApiId,
            rootResourceId: rootResourceId,
        });

        // Create API Lambda functions
        const apis: PythonLambdaFunction[] = [
            {
                name: 'list_all',
                resource: 'repository',
                description: 'List all repositories',
                path: 'repository',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'list_status',
                resource: 'repository',
                description: 'List status for all repositories',
                path: 'repository/status',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'presigned_url',
                resource: 'repository',
                description: 'Generates a presigned url for uploading files to RAG',
                path: 'repository/presigned-url',
                method: 'POST',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'create',
                resource: 'repository',
                description: 'Create a new repository',
                path: 'repository',
                method: 'POST',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'get_repository_by_id',
                resource: 'repository',
                description: 'Get a repository by ID',
                path: 'repository/{repositoryId}',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'update_repository',
                resource: 'repository',
                description: 'Update a repository',
                path: 'repository/{repositoryId}',
                method: 'PUT',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'delete',
                resource: 'repository',
                description: 'Delete a repository',
                path: 'repository/{repositoryId}',
                method: 'DELETE',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'similarity_search',
                resource: 'repository',
                description: 'Run a similarity search against the specified repository using the specified query',
                path: 'repository/{repositoryId}/similaritySearch',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'ingest_documents',
                resource: 'repository',
                description: 'Ingest a set of documents based on specified S3 path',
                path: 'repository/{repositoryId}/bulk',
                method: 'POST',
                timeout: Duration.minutes(15),
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'list_docs',
                resource: 'repository',
                description: 'List all docs for a repository',
                path: 'repository/{repositoryId}/document',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'get_document',
                resource: 'repository',
                description: 'Get a document by ID',
                path: 'repository/{repositoryId}/{documentId}',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'download_document',
                resource: 'repository',
                description: 'Creates presigned url to download document within repository',
                path: 'repository/{repositoryId}/{documentId}/download',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'delete_documents',
                resource: 'repository',
                description: 'Deletes all records associated with documents from the repository',
                path: 'repository/{repositoryId}/document',
                method: 'DELETE',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'list_jobs',
                resource: 'repository',
                description: 'List all ingestion jobs for a repository',
                path: 'repository/{repositoryId}/jobs',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'list_collections',
                resource: 'repository',
                description: 'List all collections within a repository',
                path: 'repository/{repositoryId}/collection',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'list_user_collections',
                resource: 'repository',
                description: 'List all collections user has access to across all repositories',
                path: 'repository/collections',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'create_collection',
                resource: 'repository',
                description: 'Create a new collection within a repository',
                path: 'repository/{repositoryId}/collection',
                method: 'POST',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'get_collection',
                resource: 'repository',
                description: 'Get a collection by ID within a repository',
                path: 'repository/{repositoryId}/collection/{collectionId}',
                method: 'GET',
                environment: {
                    ...baseEnvironment,
                },
            },
            {
                name: 'delete_collection',
                resource: 'repository',
                description: 'Delete a collection within a repository',
                path: 'repository/{repositoryId}/collection/{collectionId}',
                method: 'DELETE',
                environment: {
                    ...baseEnvironment,
                },
            }
        ];

        const lambdaPath = config.lambdaPath || LAMBDA_PATH;
        apis.forEach((f) => {
            const lambdaFunction = registerAPIEndpoint(
                this,
                restApi,
                lambdaPath,
                commonLayers,
                f,
                getDefaultRuntime(),
                vpc,
                securityGroups,
                authorizer,
                lambdaExecutionRole,
            );

            // Capture create_collection Lambda for backward compatibility
            if (f.name === 'create_collection') {
                this.createCollectionFunction = lambdaFunction;
            }
        });
    }
}

File: lisa-sdk/lisapy/repository.py
#   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
#   Licensed under the Apache License, Version 2.0 (the "License").
#   You may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

from typing import Dict, List

from .common import BaseMixin
from .errors import parse_error
from .types import RagRepositoryConfig


class RepositoryMixin(BaseMixin):
    """Mixin for repository-related operations."""

    def list_repositories(self) -> List[Dict]:
        """List all available repositories.

        Returns:
            List[Dict]: List of repository configurations

        Raises:
            Exception: If the request fails
        """
        response = self._session.get(f"{self.url}/repository")
        if response.status_code == 200:
            json_models: List[Dict] = response.json()
            return json_models
        else:
            raise parse_error(response.status_code, response)

    def create_repository(self, rag_config: RagRepositoryConfig) -> Dict:
        """Create a new RAG repository.

        Args:
            rag_config: Configuration for the RAG repository

        Returns:
            Dict: Created repository information

        Raises:
            Exception: If the request fails
        """
        response = self._session.post(f"{self.url}/repository", json=rag_config)
        if response.status_code in [200, 201]:
            return response.json()  # type: ignore[no-any-return]
        else:
            raise parse_error(response.status_code, response)

    def create_pgvector_repository(self, rag_config: Dict) -> Dict:
        """Create a PGVector repository configuration.

        Args:
            rag_config: RAG configuration for the PGVector repository (will be wrapped in ragConfig)

        Returns:
            Dict: Created repository information
        """
        return self.create_repository(rag_config)  # type: ignore[arg-type]

    def create_opensearch_repository(
        self,
        repository_id: str,
        repository_name: str | None = None,
        embedding_model_id: str | None = None,
        opensearch_config: Dict | None = None,
        allowed_groups: List[str] | None = None,
    ) -> Dict:
        """Create an OpenSearch repository configuration.

        Args:
            repository_id: Unique identifier for the repository
            repository_name: User-friendly name for the repository
            embedding_model_id: Default embedding model ID
            opensearch_config: OpenSearch configuration (optional - will create new cluster if not provided)
            allowed_groups: List of groups allowed access

        Returns:
            Dict: Created repository information
        """
        rag_config = {
            "repositoryId": repository_id,
            "repositoryName": repository_name or repository_id,
            "type": "opensearch",
            "embeddingModelId": embedding_model_id,
            "allowedGroups": allowed_groups or [],
        }

        if opensearch_config:
            rag_config["opensearchConfig"] = opensearch_config
        else:
            # Create new OpenSearch cluster config
            rag_config["opensearchConfig"] = {
                "dataNodes": 2,
                "dataNodeInstanceType": "r7g.large.search",
                "masterNodes": 0,
                "masterNodeInstanceType": "r7g.large.search",
                "volumeSize": 20,
                "volumeType": "gp3",
                "multiAzWithStandby": False,
            }

        return self.create_repository(rag_config)  # type: ignore[arg-type]

    def create_bedrock_kb_repository(self, rag_config: Dict) -> Dict:
        """Create a Bedrock Knowledge Base repository configuration.

        Args:
            rag_config: RAG configuration for the Bedrock KB repository

        Returns:
            Dict: Created repository information
        """
        return self.create_repository(rag_config)

    def delete_repository(self, repository_id: str) -> bool:
        """Delete a repository.

        Args:
            repository_id: The ID of the repository to delete

        Returns:
            bool: True if deletion was successful

        Raises:
            Exception: If the request fails
        """
        response = self._session.delete(f"{self.url}/repository/{repository_id}")
        if response.status_code in [200, 204]:
            return True
        else:
            raise parse_error(response.status_code, response)

    def get_repository_status(self) -> Dict:
        """Get the status of RAG repositories.

        Returns:
            Dict: Repository status information

        Raises:
            Exception: If the request fails
        """
        response = self._session.get(f"{self.url}/repository/status")
        if response.status_code == 200:
            return response.json()
        else:
            raise parse_error(response.status_code, response)
